docs/notebooks/perturbed_optimizers/perturbed_optimizers.md

```diff
@@ -0,0 +1,380 @@
+$$y^*(\theta) = \mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta \rangle$$
+
+We consider, for a random $Z$ drawn from a distribution with continuous positive distribution $\mu$
+
+$$y_\varepsilon^*(\theta) = \mathbf{E}[\mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta + \varepsilon Z \rangle]$$
+
+```{code-cell} ipython3
+:id: 9WIWwRdSU51j
+
+%%capture
+%pip install jaxopt
+```
+
+```{code-cell} ipython3
+:id: S6tLyyy9VCEw
+
+import jax
+import jax.numpy as jnp
+import jaxopt
+import time
+
+from jaxopt import perturbations
+```
+
++++ {"id": "EmYn_jNUFfw2"}
+
+# Argmax one-hot
+
++++
+
+We start from an original optimizer, an `argmax_one_hot` function, corresponding to $y^*$ for $\mathcal{C}$ being the unit simplex. We run it on an example input `values`
+
++++ {"id": "84N-wAJ8GDK2"}
+
+## One-hot function
+
+```{code-cell} ipython3
+:id: kMZnzhX4FjGj
+
+def argmax_one_hot(x, axis=-1):
+  return jax.nn.one_hot(jnp.argmax(x, axis=axis), x.shape[axis])
+```
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: iynCk8734Wiz
+outputId: 31e902b6-2dfa-4340-b11e-955b422bd313
+---
+values = jnp.array([-0.6, 1.9, -0.2, 1.1, -1.0])
+
+one_hot_vec = argmax_one_hot(values)
+print(one_hot_vec)
+```
+
++++ {"id": "6rbNt-6zGb-J"}
+
+## One-hot with pertubations
+
++++
+
+Our implementation allows to modify this function by creating a perturbed optimizer `pert_one_hot`, corresponding to $y_\varepsilon^*$ using Gumbel noise for $Z$.
+
+```{code-cell} ipython3
+:id: 7hQz6zuPwkpZ
+
+num_samples = 100000
+sigma = 0.5
+rng = jax.random.PRNGKey(1)
+pert_one_hot = perturbations.make_perturbed_argmax(argmax_fun=argmax_one_hot,
+                                         num_samples=num_samples,
+                                         sigma=sigma,
+                                         noise=perturbations.Gumbel())
+```
+
+In this particular case, it is equal to the usual softmax, with exponential weights. This is not always true, in general there is no closed form for $y_\varepsilon^*$
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: f2gDpghJYZ33
+outputId: adbe4b3c-4ec9-4f18-ea75-07619fd84cb3
+---
+rngs = jax.random.split(rng, 2)
+
+rng = rngs[0]
+
+pert_argmax = pert_one_hot(values, rng)
+print(f'computation with {num_samples} samples, sigma = {sigma}')
+print(f'perturbed argmax = {pert_argmax}')
+jax.nn.softmax(values/sigma)
+soft_max = jax.nn.softmax(values/sigma)
+print(f'softmax = {soft_max}')
+print(f'square norm of softmax = {jnp.linalg.norm(soft_max):.2e}')
+print(f'square norm of difference = {jnp.linalg.norm(pert_argmax - soft_max):.2e}')
+```
+
++++ {"id": "2U7rhtEAGpMV"}
+
+## Gradients for one-hot with perturbations
+
++++
+
+The perturbed optimizer $y_\varepsilon^*$ is differentiable, and its gradient can be computed with stochastic estimation automatically, using `jax.grad`.
+
+We create a scalar loss `loss_simplex` of the perturbed optimizer $y^*_\varepsilon$
```

Comment:
perhaps worth saying that this is a squared loss between y_true and y_eps ? This was not clear to me from the text + math